### DISCUSSION TITLE
A new toRefs, which supports converting ref.value of object type to ref.

### DISCUSSION THREAD

**Discussion by hcg1023** [vuejs/rfcs#434] [General]
### What problem does this feature solve?
In Vue3, because the object generated by reactive can not be reassigned, we generally use the method of ` const objRef = ref(object) ` to define an object that needs to be assigned repeatedly. However, such a definition will make it cumbersome to read the attributes of the object, for example:
```
interface ObjType {
  name: string
  age: number
  sex?: string
}
const objRef = ref<ObjType>({ name: 'H', age: 1 })
objRef.value.name // 'H'
objRef.value.age // 1
```
I hope to provide a quick reading method through toRefs, for example:
```
const { name, age } = toRefs(objRef)
const sex = toRef(objRef, 'sex', 'boy')
```

### What does the proposed API look like?
I have realized this function, but I have added a new api, which does not replace the original toRefs, so I hope to discuss with you whether it exists and if so, what does it look like? If appropriate, I will mention a PR. I put the corresponding code in my repository.

[https://github.com/hcg1023/vue3-dnd/blob/main/lib/internals/toReactive.ts](https://github.com/hcg1023/vue3-dnd/blob/main/lib/internals/toReactive.ts)

**Comment by posva** [comment]
This looks more like a vueuse function than a core utility

**Reply by hcg1023** [reply]
Why do you think so? toRef and toRefs are internal

**Comment by lidlanca** [comment]
```js
const { name, age } = toRefs(objRef.value)
```

**Reply by hcg1023** [reply]
That's not going to work, you're going to lose your original reactive
```ts
const { name, age } = toRefs(objRef.value)
name.value = '1'
// now
name.value !== objRef.value.name
```
I want
```ts
name.value === objRef.value.name
```

**Reply by lidlanca** [reply]
can you show how it is not working  here

https://sfc.vuejs.org/

**Reply by hcg1023** [reply]
> can you show how it is not working here
> 
> https://sfc.vuejs.org/

[https://sfc.vuejs.org/](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiwgdG9SZWZzIH0gZnJvbSAndnVlJ1xuXG4gIGNvbnN0IG9ialJlZiA9IHJlZih7XG4gICAgbmFtZTogJ3dhbmcnLFxuICAgIGFnZTogMTJcbiAgfSlcbiAgXG4gIGNvbnN0IHsgbmFtZSwgYWdlIH0gPSB0b1JlZnMob2JqUmVmLnZhbHVlKVxuICBcbiAgb2JqUmVmLnZhbHVlID0geyBuYW1lOiAneHh4JywgYWdlOiAxMyB9XG4gIC8vbm93IG9ialJlZi52YWx1ZS5uYW1lICE9PSBuYW1lLnZhbHVlXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8aDE+e3sgbmFtZSB9fTwvaDE+XG4gIDxoMT57eyBhZ2UgfX08L2gxPlxuPC90ZW1wbGF0ZT4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL3NmYy52dWVqcy5vcmcvdnVlLnJ1bnRpbWUuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59In0=)
I want to reassign objRef and make sure name and value point to my latest objectRef, which I think is the desired effect

**Reply by lidlanca** [reply]
I see, reactivity of the extracted refs can be preserved, by mutating the ref.value instead of replacing it.

[sfc](https://sfc.vuejs.org/#eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbiAgaW1wb3J0IHsgcmVmLHJlYWN0aXZlLCB0b1JlZnMsdG9SYXcsc2hhbGxvd1JlZn0gZnJvbSAndnVlJ1xuXG4gIGxldCBvID0ge1xuICAgIG5hbWU6ICdvcmlnaW5hbCcsXG4gICAgYWdlOiAxMlxuICB9XG5cbiAgbGV0IG15UmVmID0gcmVmKG8pXG5cbiAgY29uc3QgeyBuYW1lLCBhZ2UgfSA9IHRvUmVmcyhteVJlZi52YWx1ZSlcblxuICBmdW5jdGlvbiBzZXRSZWZUcmFja2FibGUocixvKXtcbiAgICBsZXQgb2sgPSBPYmplY3Qua2V5cyhvKTtcbiAgICBmb3IobGV0IGsgaW4gci52YWx1ZSkgXG4gICAgICBvay5pbmNsdWRlcyhrKSA/IHIudmFsdWVba10gPSBvW2tdIDogZGVsZXRlIHIudmFsdWVba107XG4gIH1cblxuICBmdW5jdGlvbiBkb1JlcGxhY2UoKXtcbiAgICBzZXRSZWZUcmFja2FibGUobXlSZWYsIHtuYW1lOiduZXcgbmFtZSd9KVxuICB9XG5cblxuPC9zY3JpcHQ+XG5cbjx0ZW1wbGF0ZT5cbiAgPGgxPm5hbWU6IHt7IG5hbWUgfX08L2gxPlxuICA8aDE+YWdlOiAge3sgYWdlIH19PC9oMT5cblxuICA8YnI+XG4gIDxidXR0b24gQGNsaWNrPVwiZG9SZXBsYWNlXCI+UmVwbGFjZTwvYnV0dG9uPjxicj5cbiAgPGlucHV0IHYtbW9kZWw9bmFtZT48YnI+XG4gIDxpbnB1dCB2LW1vZGVsPWFnZT48YnI+XG4gIDxicj5cbiAge3tteVJlZn19XG48L3RlbXBsYXRlPiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy92dWUucnVudGltZS5lc20tYnJvd3Nlci5qc1wiXG4gIH1cbn0ifQ==)

**Reply by hcg1023** [reply]
> 

I understand it can be done, but from a usage point of view, `objRef.value = { name: 'new' }` is more convenient

**Comment by hcg1023** [comment]
[https://github.com/hcg1023/vue-ref2reactive](https://github.com/hcg1023/vue-ref2reactive)
This is my implementation, I hope you give a proper name, if it can be built into Vue, I can also mention a PR

**Comment by hcg1023** [comment]
Can we add a new overload to toRef and toRefs to support multi-level progression of objects
e.g.
```ts
function toRef<T extends object, K extends keyof T, K2 extends keyof T[K]>(obj: T, keys: [K, K2?]): K2 extends keyof T[K] ? T[K][K2] : T[K] {
}
const nameRef = toRef({ value: { name: 'new' } }, ['value', 'name'])
nameRef.value // 'new'

const objRef = ref({ name: 'ref', age: 16 })
const { name, age } = toRefs(objRef, ['value'])
name.value // 'ref'
age.value // 16
```
